### PR TITLE
[Feature] Jenkins configuration-as-code

### PR DESCRIPTION
**Pull request by ivanchimeno**
### Motivation and Context

Setting Jenkins and getting it to run with Artemis is time consuming due to the many manual steps needed (install plugins, create credentials, etc).

### Description

This pr introduces a couple of quality-of-life changes that help speed up the setup process. These are the features:

1. Update the Jenkins dockerfile to automatically install the Jenkins plugins required by Artemis. The plugins are maintained in `src/main/docker/jenkins/plugin.yml`
2. A new `jenkins-casc-config.yml` file: We take advantage of [JCasC](https://github.com/jenkinsci/configuration-as-code-plugin) to provide a file that pre-configures the Jenkins environment.
3. A new `gitlab-jenkins-mysql.yml` docker-compose file which deploys all three environments into containers. Here we configure the jenkins service so that it automatically reads `jenkins-casc-config.yml` on startup.
4. Update the docs with a Jenkins quick-start guide.

### Test Coverage
Nothing changed

### CODE REVIEW

File: docs/dev/setup/jenkins-gitlab.rst [hunk 1]
```diff
@@ -356,8 +356,89 @@ You can also remove all old images using ``docker image prune -a``
...
+                password: artemis-admin
+            version-control:
+                url: http://172.33.0.2:8081
+                user: artemis_admin
+                password: artemis_admin
+                ci-token: # generated in step 9
+            continuous-integration:
+                url: http://172.33.0.3:8082
+                user: artemis_admin
+                password artemis_admin
```
Comment (b-fein) [inline docs/dev/setup/jenkins-gitlab.rst:432]:
```suggestion
                password: artemis_admin
```

File: docs/dev/setup/jenkins-gitlab.rst [hunk 2]
```diff
@@ -356,8 +356,89 @@ You can also remove all old images using ``docker image prune -a``
...
+                vcs-credentials: artemis_gitlab_admin_credentials
+                artemis-authentication-token-key: artemis_notification_plugin_token
+                artemis-authentication-token-value: artemis_admin
+                secret-push-token: # generated in step 9
```
Comment (b-fein) [inline docs/dev/setup/jenkins-gitlab.rst:436]:
```suggestion
                secret-push-token: # generated in step 8
```

File: docs/dev/setup/jenkins-gitlab.rst [hunk 3]
```diff
@@ -356,8 +356,89 @@ You can also remove all old images using ``docker image prune -a``
...
+.. code:: yaml
+
+    artemis:
+        user-management:
+            use-external: false
+            internal-admin:
+                username: artemis_admin
+                password: artemis-admin
+            version-control:
+                url: http://172.33.0.2:8081
```
Comment (b-fein) [inline docs/dev/setup/jenkins-gitlab.rst:425]:
```suggestion
                url: http://172.33.0.2
```

Gitlab listens on `172.33.0.2:80` on the internal Docker-network. `8081` is only the external port that is forwarded to that internal IP. Accessing Gitlab with port `8081` only works for `http://localhost:8081` for me.

Also Artemis throws a `ConnectionFailureException` when trying it with `8081`.

File: docs/dev/setup/jenkins-gitlab.rst [hunk 4]
```diff
@@ -356,8 +356,89 @@ You can also remove all old images using ``docker image prune -a``
...
+                user: artemis_admin
+                password: artemis_admin
+                ci-token: # generated in step 9
+            continuous-integration:
+                url: http://172.33.0.3:8082
```
Comment (b-fein) [inline docs/dev/setup/jenkins-gitlab.rst:430]:
```suggestion
                url: http://172.33.0.3:8080
```

Same as with Gitlab above.

File: src/main/docker/gitlab-jenkins-mysql.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        hostname: jenkins
+        environment:
+            - CASC_JENKINS_CONFIG=/usr/share/jenkins/ref/jenkins-casc-config.yml
+        networks:
+            artemis:
+                ipv4_address: 172.33.0.3
+    mysql:
+        command: mysqld --lower_case_table_names=1 --skip-ssl --character_set_server=utf8mb4 --explicit_defaults_for_timestamp
+        environment:
+            - MYSQL_USER=root
```
Comment (b-fein) [inline src/main/docker/gitlab-jenkins-mysql.yml:35]:
```suggestion
```

This line has to be removed. The default MySQL user is already `root` and MySQL refuses to start up when the variable is set explicitly to `root`. `docker logs docker_mysql_1` when the variable is set explains the reasoning in more detail.

File: src/main/docker/gitlab-jenkins-mysql.yml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            - MYSQL_ALLOW_EMPTY_PASSWORD=yes
+            - MYSQL_DATABASE=Artemis
+        image: mysql:8.0.23
+        ports:
+            - 3306:3306
+        volumes:
+            - ./data/.db:/var/lib/mysql
```
Comment (b-fein) [inline src/main/docker/gitlab-jenkins-mysql.yml:42]: Is there a reason besides backwards-compatibility why the MySQL database is mapped to a local directory and not a Docker volume like it is done with Gitlab and Jenkins data?

File: src/main/docker/jenkins/jenkins-casc-config.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            - "Overall/Administer:artemis_admin"  # Make sure that the username matches the Artemis admin username
+            - "Overall/Read:authenticated"
+credentials:
+  system:
+    domainCredentials:
+      - credentials:
+          - gitLabApiTokenImpl:
+              id: artemis_gitlab_api_token
+              scope: GLOBAL
+              apiToken: c3jyyTTyJ3pmyPJEq3ML # Access token generated in Gitlab with the "api" and "read_repository" permissions.
```
Comment (b-fein) [inline src/main/docker/jenkins/jenkins-casc-config.yml:22]:
```suggestion
              apiToken: your.gitlab.api.token # Access token generated in Gitlab with the "api" and "read_repository" permissions.
```

**Review by b-fein — COMMENTED** [review]
This makes the local development setup a lot easier. GJ!

I had to make a few changes to the configuration to get all containers started as intended.

The Jenkins Plugins seem to get installed properly and Build-Jobs start, so the CASC-setup seems to generally work.
I need to retest it again, though, as at the moment setting up the Webhooks in Gitlab fails (‘Invalid URL’) and Jenkins couldn’t pull code from Gitlab. I assume that is just my fault by having old API-Tokens somewhere that were only valid in some other older containers.

File: src/main/docker/gitlab-jenkins-mysql.yml (hunk 5 above)
Comment (ivanchimeno) [reply src/main/docker/gitlab-jenkins-mysql.yml:35]: Thanks for the comment. I tested it and indeed we don't need that entry anymore.

File: docs/dev/setup/jenkins-gitlab.rst (hunk 3 above)
Comment (ivanchimeno) [reply docs/dev/setup/jenkins-gitlab.rst:425]: Good catch!

File: src/main/docker/gitlab-jenkins-mysql.yml (hunk 6 above)
Comment (b-fein) [reply src/main/docker/gitlab-jenkins-mysql.yml:41]: Not having a `data` directory owned by root in the top level of the development directory seems like a good choice to me.
The only downside I can see is that users changing from a similar previous setup might have to start with a fresh database, copy the old one over into the volume, or manually change this entry back to the current state.

> Would it be better to change that line into - artemis-mysql-data:/var/lib/mysql?

I think so, yes. That change and adding `artemis-mysql-data` to the list of volumes at the bottom of the file should be enough to make it work.

**Comment by ivanchimeno** [comment]
> This makes the local development setup a lot easier. GJ!
> 
> I had to make a few changes to the configuration to get all containers started as intended.
> 
> The Jenkins Plugins seem to get installed properly and Build-Jobs start, so the CASC-setup seems to generally work.
> I need to retest it again, though, as at the moment setting up the Webhooks in Gitlab fails (‘Invalid URL’) and Jenkins couldn’t pull code from Gitlab. I assume that is just my fault by having old API-Tokens somewhere that were only valid in some other older containers.

Thank you @b-fein for your review! Are you running the local environment on a Mac or on a Linux machine? I could try to re-deploy everything and see if I encounter the same problems.

**Comment by b-fein** [comment]
> Are you running the local environment on a Mac or on a Linux machine?

It’s on Linux, with no special system-wide Docker/networking configurations as far as I know. (Docker version 20.10.5-ce, docker-compose version 1.28.5)

**Comment by ivanchimeno** [comment]
> > This makes the local development setup a lot easier. GJ!
> > I had to make a few changes to the configuration to get all containers started as intended.
> > The Jenkins Plugins seem to get installed properly and Build-Jobs start, so the CASC-setup seems to generally work.
> > I need to retest it again, though, as at the moment setting up the Webhooks in Gitlab fails (‘Invalid URL’) and Jenkins couldn’t pull code from Gitlab. I assume that is just my fault by having old API-Tokens somewhere that were only valid in some other older containers.
> 
> Thank you @b-fein for your review! Are you running the local environment on a Mac or on a Linux machine? I could try to re-deploy everything and see if I encounter the same problems.

> > Are you running the local environment on a Mac or on a Linux machine?
> 
> It’s on Linux, with no special system-wide Docker/networking configurations as far as I know. (Docker version 20.10.5-ce, docker-compose version 1.28.5)

Did you check the `spring.server.url` entry inside the `application-dev.yml` file? I had to add the external Ip of the Artemis server (it's displayed when it's started) including the port. For some reason the port isn't taken from the `spring.server.port` entry.

**Comment by b-fein** [comment]
I actually had
```yaml
server:
    port: 8080
    url: http://192.168.88.218:8080
```
as part of `application-local.yml` before. Changing `url` to `url: http://172.33.0.1:8080` the IP inside the Docker network works.

**Review by b-fein — APPROVED** [review]
The pipelines need the `cleanWs` task which is provided by the Workspace Cleanup plugin.
Adding `- artifactId: ws-cleanup` to `src/main/docker/jenkins/plugins.yml` fixes failing pipelines.

After adding that change I tested setting up various exercises with Java (+ optional SCA), Swift, and Python. Repos were created, build jobs ran, and results were delivered to Artemis. Additionally I tested creating users which were successfully created in Gitlab+Jenkins, too.

The communication between the different containers and Artemis therefore worked as intended.
With the remaining little change to `plugins.yml` I approve!

**Comment by ivanchimeno** [comment]
> The pipelines need the `cleanWs` task which is provided by the Workspace Cleanup plugin.
> Adding `- artifactId: ws-cleanup` to `src/main/docker/jenkins/plugins.yml` fixes failing pipelines.
> 
> After adding that change I tested setting up various exercises with Java (+ optional SCA), Swift, and Python. Repos were created, build jobs ran, and results were delivered to Artemis. Additionally I tested creating users which were successfully created in Gitlab+Jenkins, too.
> 
> The communication between the different containers and Artemis therefore worked as intended.
> With the remaining little change to `plugins.yml` I approve!

Thanks for testing out the changes! I committed the missing plugin and the mysql volume.

**Review by b-fein — DISMISSED** [review]
Looks good! Re-approving.

**Review by FuchsDominik — APPROVED** [review]
So text-wise it looks good, only found one minor remark :)

File: docs/dev/setup/jenkins-gitlab.rst [hunk 8]
```diff
@@ -356,8 +356,89 @@ You can also remove all old images using ``docker image prune -a``
...
+
+7. You can now deploy Jenkins. A ``src/main/docker/gitlab-jenkins-mysql.yml`` file is provided which deploys the Jenkins, Gitlab, and Mysql containers bound to static ip addresses. You can deploy them by running:
+
+::
+
+       docker-compose -f src/main/docker/gitlab-jenkins-mysql.yml up --build
+
+If you already have a Gitlab and Mysql instance running, you can comment out all services except for jenkins and then run the docker-compose file.
+
+8. You need to generate the `ci-token` and `secret-push-token`. Please follow the `Gitlab to Jenkins push notification token <##gitlab-to-jenkins-push-notification-token>`__. steps
```
Comment (FrankeLukas) [inline docs/dev/setup/jenkins-gitlab.rst:412]:
```suggestion
8. You need to generate the `ci-token` and `secret-push-token`. Please follow the `Gitlab to Jenkins push notification token <##gitlab-to-jenkins-push-notification-token>`__ steps.
```

**Review by FrankeLukas — APPROVED** [review]
Changes lgtm otherwise.

**Review by FrankeLukas — APPROVED** [review]
Re-approve.

**Review by b-fein — APPROVED** [review]
Re-approve

**Review by FuchsDominik — APPROVED** [review]

**Review by derLalla — APPROVED** [review]
the changes look good to me!

**Pull request merged**